v3/newrelic/log_event.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2020 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package newrelic
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// Exported Constants for log decorators
+const (
+	// LogSeverityFieldName is the name of the log level field in New Relic logging JSON
+	LogSeverityFieldName = "level"
+
+	// LogMessageFieldName is the name of the log message field in New Relic logging JSON
+	LogMessageFieldName = "message"
+
+	// LogTimestampFieldName is the name of the timestamp field in New Relic logging JSON
+	LogTimestampFieldName = "timestamp"
+
+	// LogSpanIDFieldName is the name of the span ID field in the New Relic logging JSON
+	LogSpanIDFieldName = "span.id"
+
+	// LogTraceIDFieldName is the name of the trace ID field in the New Relic logging JSON
+	LogTraceIDFieldName = "trace.id"
+
+	// LogSeverityUnknown is the value the log severity should be set to if no log severity is known
+	LogSeverityUnknown = "UNKNOWN"
+
+	// MaxLogLength is the maximum number of bytes the log message is allowed to be
+	MaxLogLength = 32768
+)
+
+// internal variable names and constants
+const (
+	// number of bytes expected to be needed for the average log message
+	averageLogSizeEstimate = 400
+)
+
+type logEvent struct {
+	priority  priority
+	timestamp int64
+	severity  string
+	message   string
+	spanID    string
+	traceID   string
+}
+
+// LogData contains data fields that are needed to generate log events.
+type LogData struct {
+	Timestamp int64           // Optional: Unix Millisecond Timestamp; A timestamp will be generated if unset
+	Severity  string          // Optional: Severity of log being consumed
+	Message   string          // Optional: Message of log being consumed; Maximum size: 32768 Bytes.
+	Context   context.Context // Optional: context containing a New Relic Transaction
+}
+
+// writeJSON prepares JSON in the format expected by the collector.
+func (e *logEvent) WriteJSON(buf *bytes.Buffer) {
+	w := jsonFieldsWriter{buf: buf}
+	buf.WriteByte('{')
+	w.stringField(LogSeverityFieldName, e.severity)
+	w.stringField(LogMessageFieldName, e.message)
+
+	if len(e.spanID) > 0 {
+		w.stringField(LogSpanIDFieldName, e.spanID)
+	}
+	if len(e.traceID) > 0 {
+		w.stringField(LogTraceIDFieldName, e.traceID)
+	}
+
+	w.needsComma = false
+	buf.WriteByte(',')
+	w.intField(LogTimestampFieldName, e.timestamp)
+	buf.WriteByte('}')
+}
+
+// MarshalJSON is used for testing.
+func (e *logEvent) MarshalJSON() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, averageLogSizeEstimate))
```

Comment:
Does hinting this with that "400" really make that much of a difference? Cool that we're thinking this hard about performance, though. I'd love to hear about your grouping of `int`s that I read earlier, too.

Reply:
Yes! The bytes buffer gets allocated on the heap, and when it grows, it has to estimate/request more memory. Each time it does that, we incur a cpu cost. By estimating a little more closely to what we think we need, we can cut down the number of buffer resizes we need, and save us cpu time, and run time!

Reply:
Yes, it's similar to C in that respect.